presto-jmx/src/main/java/com/facebook/presto/connector/jmx/JmxRecordSetProvider.java

```diff
@@ -62,116 +67,154 @@ public RecordSet getRecordSet(ConnectorTransactionHandle transaction, ConnectorS
+            }
+        }
+        catch (JMException e) {
+            rows = ImmutableList.of();
+        }
+
+        return new InMemoryRecordSet(columnTypes.values(), rows);
+    }
+
+    private List<Integer> calculateSelectedColumns(List<JmxColumnHandle> columnHandles, Map<String, Type> selectedColumnsTypes)
+    {
+        ImmutableList.Builder<Integer> selectedColumns = ImmutableList.builder();
+        for (int i = 0; i < columnHandles.size(); i++) {
+            JmxColumnHandle column = columnHandles.get(i);
+            if (selectedColumnsTypes.containsKey(column.getColumnName())) {
+                selectedColumns.add(i);
+            }
+        }
+        return selectedColumns.build();
+    }
+
+    public ImmutableMap<String, Type> getColumnTypes(List<? extends ColumnHandle> columnHandles)
+    {
         ImmutableMap.Builder<String, Type> builder = ImmutableMap.builder();
-        for (ColumnHandle column : columns) {
+        for (ColumnHandle column : columnHandles) {
             JmxColumnHandle jmxColumnHandle = checkType(column, JmxColumnHandle.class, "column");
             builder.put(jmxColumnHandle.getColumnName(), jmxColumnHandle.getColumnType());
         }
-        ImmutableMap<String, Type> columnTypes = builder.build();
+        return builder.build();
+    }
 
-        List<List<Object>> rows;
-        try {
-            Map<String, Object> attributes = getAttributes(columnTypes.keySet(), tableHandle);
-            List<Object> row = new ArrayList<>();
-            // NOTE: data must be produced in the order of the columns parameter.  This code relies on the
-            // fact that columnTypes is an ImmutableMap which is an order preserving LinkedHashMap under
-            // the covers.
-            for (Entry<String, Type> entry : columnTypes.entrySet()) {
-                if (entry.getKey().equals("node")) {
-                    row.add(nodeId);
+    private ImmutableMap<String, Optional<Object>> getAttributes(Set<String> uniqueColumnNames, JmxTableHandle tableHandle)
+            throws JMException
+    {
+        ObjectName objectName = new ObjectName(tableHandle.getObjectName());
+
+        String[] columnNamesArray = uniqueColumnNames.toArray(new String[uniqueColumnNames.size()]);
+
+        ImmutableMap.Builder<String, Optional<Object>> attributes = ImmutableMap.builder();
+        for (Attribute attribute : mbeanServer.getAttributes(objectName, columnNamesArray).asList()) {
+            attributes.put(attribute.getName(), Optional.ofNullable(attribute.getValue()));
+        }
+        return attributes.build();
+    }
+
+    public List<Object> getLiveRow(JmxTableHandle tableHandle, ImmutableMap<String, Type> columnTypes)
+            throws JMException
+    {
+        return getLiveRow(tableHandle, columnTypes, 0);
+    }
+
+    public List<Object> getLiveRow(JmxTableHandle tableHandle, ImmutableMap<String, Type> columnTypes, long dumpTimestamp)
```

Comment:
`dumpTimestamp` is a weird name. I'd just call it `timestamp` or `entryTimestamp`